packages/engine/src/avatar/AvatarInputSystem.ts

```diff
@@ -0,0 +1,180 @@
+import { useEffect } from 'react'
+import { Vector3 } from 'three'
+
+import { isDev } from '@xrengine/common/src/config'
+import { dispatchAction, getState, startReactor, useHookstate } from '@xrengine/hyperflux'
+
+import { EngineActions } from '../ecs/classes/EngineState'
+import { World } from '../ecs/classes/World'
+import { getComponent } from '../ecs/functions/ComponentFunctions'
+import { ButtonInputState } from '../input/InputState'
+import { InteractState } from '../interaction/systems/InteractiveSystem'
+import { WorldNetworkAction } from '../networking/functions/WorldNetworkAction'
+import { boxDynamicConfig } from '../physics/functions/physicsObjectDebugFunctions'
+import { accessEngineRendererState, EngineRendererAction } from '../renderer/EngineRendererState'
+import { getControlMode } from '../xr/XRState'
+import { AvatarControllerComponent } from './components/AvatarControllerComponent'
+import { moveAvatarWithTeleport, rotateAvatar } from './functions/moveAvatar'
+import { AvatarInputSettingsState } from './state/AvatarInputSettingsState'
+
+export default async function AvatarInputSystem(world: World) {
+  const keyState = getState(ButtonInputState)
+  const interactState = getState(InteractState)
+  const avatarInputSettingsState = getState(AvatarInputSettingsState)
+
+  const reactor = startReactor(() => {
+    const keys = useHookstate(keyState)
+
+    useEffect(() => {
+      if (keys.value.ShiftLeft && world.localClientEntity) {
+        const controller = getComponent(world.localClientEntity, AvatarControllerComponent)
+        controller.isWalking = !controller.isWalking
+      }
+    }, [keys.ShiftLeft])
+
+    useEffect(() => {
+      if (keys.value.KeyE && world.localClientEntity) {
```

Comment:
I am a little bit worried about turning input events into an input state that is handled reactively, and thus potentially delayed. I think this means it's possible, for example, to press the E key, and then by the time that this useEffect runs, the E key will no longer be pressed—meaning nothing happens.